### PR TITLE
feat: Allow installation of Kind directly from extension

### LINKED ISSUE #1257: Kind: Allow installation of Kind directly from extension
### Is your feature request related to a problem? Please describe

As a developer, I would like to be able to get the flow to install Kind directly from the extension, so I can remain in the application to setup a kubernetes local cluster.

### PR DESCRIPTION
**Pull request by jeffmaury**
Fixes #1257

Signed-off-by: Jeff MAURY <jmaury@redhat.com>

### What does this PR do?

Kind extension is checking if kind is available from PATH and if not propose the install from GitHub.

### Screenshot/screencast of this PR

kind

## Status bar implementation

kind

### What issues does this PR fix or reference?

Fixes #1257

### CODE REVIEW

File: extensions/kind/src/extension.ts [hunk 1]
```diff
@@ -99,7 +102,7 @@ nodes:
     await fs.promises.writeFile(tmpFilePath, kindClusterConfig, 'utf8');
 
     // now execute the command to create the cluster
-    const result = await runCliCommand('kind', ['create', 'cluster', '--config', tmpFilePath], { env, logger });
+    const result = await runCliCommand(kindCli, ['create', 'cluster', '--config', tmpFilePath], { env, logger });
```
Comment (lstocchi) [inline extensions/kind/src/extension.ts:108]: Could this be undefined? Does it need a check?

**Review by lstocchi — APPROVED** [review]
LGTM. Works fine on Windows

File: extensions/kind/package.json [hunk 2]
```diff
@@ -51,7 +51,9 @@
     "watch": "tsc -w"
   },
   "dependencies": {
-    "@tmpwip/extension-api": "^0.0.1"
+    "@tmpwip/extension-api": "^0.0.1",
+    "@terascope/fetch-github-release": "^0.8.7",
```
Comment (benoitf) [inline extensions/kind/package.json]: I would avoid this dependency if possible. It seems to depend on too many extra dependencies 

Also when checking PR with extensions, please also check a production binary

**Review by benoitf — COMMENTED** [review]
I'm not sure that it should be a considered as a 'main provider 'being present on all the pages. It seems it's more something that should live in the 'status bar'

Not sure how many people would understand what is kind and why they need to install it.

File: extensions/kind/src/kind-installer.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+const githubOrganization = 'kubernetes-sigs';
+const githubRepo = 'kind';
+
+export class KindInstaller {
+  private assetNames = new Map<string, string>();
+
+  private assetPromise: Promise<string>;
+
+  constructor(private readonly storagePath: string) {
+    this.assetNames.set('win32-x64', 'kind-windows-amd64');
```
Comment (cdrage) [inline extensions/kind/src/kind-installer.ts]: should we make these const variables?

**Review by cdrage — DISMISSED** [review]

File: extensions/kind/scripts/build.js [hunk 4]
```diff
@@ -34,3 +36,6 @@ if (fs.existsSync(builtinDirectory)) {
 }
 
 zipper.sync.zip(path.resolve(__dirname, '../')).compress().save(destFile);
+mkdirp(unzippedDirectory).then(() => {
```
Comment (benoitf) [inline extensions/kind/scripts/build.js]: this part will enable kind extension by default, so I'm unsure we want to have it enabled by default yet
cc @slemeur

File: extensions/kind/src/extension.ts [hunk 5]
```diff
@@ -222,6 +220,33 @@ export async function activate(extensionContext: extensionApi.ExtensionContext):
...
+
+export async function activate(extensionContext: extensionApi.ExtensionContext): Promise<void> {
+  const installer = new KindInstaller(extensionContext.storagePath);
+  kindCli = await detectKind(extensionContext.storagePath, installer);
+
+  if (!kindCli) {
+    if (await installer.isAvailable()) {
+      const statusBarItem = extensionApi.window.createStatusBarItem();
+      statusBarItem.text = 'Kind';
+      statusBarItem.tooltip = 'Kind not found on your system, click to download and install it';
```
Comment (benoitf) [inline extensions/kind/src/extension.ts:235]: it's missing probably
```
statusBarItem.iconClass = 'fa fa-exclamation-triangle';
```

to notify that there is a problem on Kind like 
image

File: extensions/kind/src/kind-installer.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+    return this.assetPromise;
+  }
+
+  async isAvailable(): Promise<boolean> {
+    const assetInfo = await this.getAssetInfo();
+    return assetInfo !== undefined;
+  }
+
+  async performInstall(): Promise<void> {
+    console.log('Installing kind');
```
Comment (benoitf) [inline extensions/kind/src/kind-installer.ts]: we can get rid of this log

File: extensions/kind/src/kind-installer.ts [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+    const dialogResult = await extensionApi.window.showInformationMessage(
+      `kind is not installed on this system, would you like to install Kind ?`,
```
Comment (benoitf) [inline extensions/kind/src/kind-installer.ts]: there is linter error on Github: Strings must use singlequote

**Review by benoitf — CHANGES_REQUESTED** [review]
if I click on kind icon and like do a mistake and click on no
the icon is gone


https://user-images.githubusercontent.com/436777/218440529-c548d3a9-632b-42c6-aae6-d7ede305126f.mp4


I would expect the warning be there until I click yes

(or a configuration property to not bother me again)

**Comment by benoitf** [comment]
if I click 'no', now I get in the logs

```
StatusBarItem.svelte? [sm]:39 Uncaught (in promise) Error
```

**Comment by benoitf** [comment]
and I click on yes I've

```
StatusBarItem.svelte? [sm]:39 Uncaught (in promise) Error: ENOENT: no such file or directory, open '/Users/benoitf/.podman-desktop/kind/kind-darwin-amd64'
```

**Comment by jeffmaury** [comment]
> and I click on yes I've
> 
> ```
> StatusBarItem.svelte? [sm]:39 Uncaught (in promise) Error: ENOENT: no such file or directory, open '/Users/benoitf/.podman-desktop/kind/kind-darwin-amd64'
> ```

Can reproduce on Windows. Can you tell me what's in $HOME/.podman-desktop/kind ?

File: extensions/kind/scripts/build.js (hunk 4 above)
Comment (jeffmaury) [reply extensions/kind/scripts/build.js]: Reverted the commit

**Comment by benoitf** [comment]
```sh
ls $HOME/.podman-desktop/kind 
ls: /Users/benoitf/.podman-desktop/kind: No such file or directory
```

```sh
ls $HOME/.podman-desktop/
podman
```

**Comment by benoitf** [comment]
it seems to download the kind binary but I'm not notified it has downloaded the file and the permission is not set
so it looks it's somehow blocked somewhere

https://user-images.githubusercontent.com/436777/218967054-ce38fcdd-bc34-4ca7-a6ec-0de9f523e240.mp4

**Comment by jeffmaury** [comment]
> it seems to download the kind binary but I'm not notified it has downloaded the file and the permission is not set so it looks it's somehow blocked somewhere

Will add an error message if something is wrong during the installation

**Comment by benoitf** [comment]
image

**Review by benoitf — APPROVED** [review]
kind is now successfully downloaded.

But:
- I would expect a 'succeed notification'. I clicked on the warning icon, it downloaded kind, the status bar element disappeared so I guessed it succeed but I would have expected a notification that it was successful.
- While kind is downloaded, I can't use it from a CLI (so I don't know if we should ask the user to add a specific PATH etc)

**Comment by jeffmaury** [comment]
> kind is now successfully downloaded.
> 
> But:
> 
>     * I would expect a 'succeed notification'. I clicked on the warning icon, it downloaded kind, the status bar element disappeared so I guessed it succeed but I would have expected a notification that it was successful.
> 
>     * While kind is downloaded, I can't use it from a CLI (so I don't know if we should ask the user to add a specific PATH etc)

There is a notification, I can see it on Windows. Do we have an issue on MacOS.
Will open another issue for PATH

**Comment by benoitf** [comment]
here is a recording on what I have
binary looks like to be successfully downloaded but I don't ave the notification that it's successful. Status bar just disappear.

https://user-images.githubusercontent.com/436777/220243917-2e830ee1-5b75-4562-bf58-9d8aba34530e.mp4

**Comment by jeffmaury** [comment]
> here is a recording on what I have binary looks like to be successfully downloaded but I don't ave the notification that it's successful. Status bar just disappear.

`showNotification` is called. Maybe let's ask another MacOS dev to test it @cdrage ?

**Comment by cdrage** [comment]
https://user-images.githubusercontent.com/6422176/220490798-f8668c97-1999-43b8-bff1-7a841f406738.mov

When clicking no I expect the notification to go away, especially since I do not use `kind` locally (I have an external kube-context). This might be annoying for other users.

**Comment by cdrage** [comment]
When testing, I tried installing and clicking "yes". All it did is make the notification disappear and then there we no other notifications?

**Comment by jeffmaury** [comment]
> When testing, I tried installing and clicking "yes". All it did is make the notification disappear and then there we no other notifications?

Seems to be an issue on MacOS as I am seeing the notification on Windows. Will ask another check on Windows.

**Comment by lstocchi** [comment]
Tried on both Windows 10 and Fedora 37, i can see the notification. The issue seems to be only on MacOS

**Comment by benoitf** [comment]
As we're working on the Notification Center, let's merge.
It could have been nice to have a test that check that the notifications are called (or the error is notified upon error)

**Pull request merged**